[Builder] Optional YAML based build configuration flow

bwintermann opened the pull request:
Draft for a YAML based build file, replacing the need for Python scripts, but not replacing it (keeping backwards compatibility, so writing a `build.py` script is still possible in case your build flow is more complicated). Would work on this occasionally unless @iksnagreb  already has something similar.

I think it would be cleaner and presents a nice interface with software written around it which does not necessarily use Python.

Tasks that would need to be done:

- [x] YAML to BuildDataflowCfg conversion function
- [x] Integration into `run-docker.sh` so that the correct FINN function is called (ideally you can pass both `build.py` and `build.yaml` to `./run-docker.sh build_custom` and the script automatically can differentiate)
- [x] Implement every option that BuildDataflowCfg has to be settable by the YAML file
- [ ] Implement type checking of imported YAML arguments - atleast for the most important arguments. Still have to figure out how to do it efficiently, reading `__annotations__` via string comparison seems very unstable
- [ ] Implement setup and teardown hooks for scripts for custom functionality
- [ ] ~Replace print statements by logger statements when #14 is merged~ (Probably won't be possible)
- [x] Tests for the conversion function and helper functions
- [ ] 100% Test coverage and confirmed working status

An example `build.yaml` could look like this:
```YAML
config:
  general:
    synth_clk_period_ns: 5.5
    board: U280
    platform: Alveo
    vitis_opt_strategy: PERFORMANCE
    specialize_layers_config_file: specialization.json
    folding_config_file: folding.json
    generate_outputs:
        - BITFILE
        - DEPLOYMENT_PACKAGE
        - ...
  
  build_steps:
    - step_convert_qonnx
    - step_streamline
    - step_create_dataflow_partition
    - step_specialize_layers
    - step_generate_estimate_reports
    - step_hw_codegen
    - step_hw_ipgen
    - custom_steps.step_my_great_function
    - ...
```

iksnagreb commented:
> unless @iksnagreb already has something similar.

I actually kind of already have, except for the build_steps as these sometimes require stuff computed/inserted/configured from other sources which is not easy to do with static YAML configuration (some of my custom build steps are actually build step generators...). However, the way I currently handle the YAML-based config is to simply `**config` into the `DataflowBuildConfig`, so there is probably something to refine... I will have a look at your draft, maybe we could/should discuss this further. I am also not sure what @fpjentzsch might already have in place in terms of JSON-based infrastructure...

iksnagreb commented:
> unless @iksnagreb already has something similar.

I actually kind of already have, except for the build_steps as these sometimes require stuff computed/inserted/configured from other sources which is not easy to do with static YAML configuration (some of my custom build steps are actually build step generators...). However, the way I currently handle the YAML-based config is to simply `**config` into the `DataflowBuildConfig`, so there is probably something to refine... I will have a look at your draft, maybe we could/should discuss this further. I am also not sure what @fpjentzsch might already have in place in terms of JSON-based infrastructure...

bwintermann commented:
Sounds great, lets maybe discuss this futher next week, maybe we can make it flexible enough to even allow build steps generator in the YAML itself. I think in terms of JSON this should be easy to integrate since we can quite easily convert the two formats into each other.

bwintermann commented:
This is now mostly in a merge ready state. Tomorrow we can discuss what else would be missing. I'd clean up everything once and add some more tests as well. (As well as additional features that we might still need).

bwintermann commented:
Reverted changes to `build_dataflow.py` and `run_docker.sh`. Tests still pass, added the function as a static method to the `DataflowBuildConfig`. We might be able to remove some checks in the conversion, since you mentioned that now all arguments to `DataflowBuildConfig` are optional. @fpjentzsch @iksnagreb @LinusJungemann  feel free to read over it and maybe also run the tests on your setups to check if everything is alright so that we can merge quickly.

bwintermann commented:
Removed `dataclasses_json` and added `mashumaro[yaml]` to requirements and added both mixins to DataflowBuildConfig, so that we can use `DataflowBuildConfig.from_yaml()` and `DataflowBuildConfig.from_json()` to create DataflowBuildConfig instances. 

Things still to do would be checking validity of the parameters and inclusion of custom steps, but this depends on #25 and since we support the old Python-script based flow anyways, I would add these changes later.
